stdlib/dynarray.ml

```diff
@@ -0,0 +1,752 @@
+     | Elem s -> s.v <- x
+]}
+   With this approach, accessing an element still pays the cost of an
+   extra indirection (compared to approaches that do not box elements
+   in the backing array), but only operations that add new elements at
+   the end of the array pay extra allocations.
+
+   There are some situations where ['a option] is better: it makes
+   [pop_last_opt] more efficient as the underlying option can be
+   returned directly, and it also lets us use [Array.blit] to
+   implement [append]. We believe that optimzing [get] and [set] is
+   more important for dynamic arrays.
+
+   {2 Invariants and valid states}
+
+   We enforce the invariant that [length >= 0] at all times.
+   we rely on this invariant for optimization.
+
+   The following conditions define what we call a "valid" dynarray:
+   - valid length: [length <= Array.length arr]
+   - no missing element in the live space:
+     forall i, [0 <= i <=length] implies [arr.(i) <> Empty]
+   - no element in the empty space:
+     forall i, [0 <= i < length] implies [arr.(i) = Empty]
+
+   Unfortunately, we cannot easily enforce validity as an invariant in
+   presence of concurrent udpates. We can thus observe dynarrays in
+   "invalid states". Our implementation may raise exceptions or return
+   incorrect results on observing invalid states, but of course it
+   must preserve memory safety.
+*)
+
+module Error = struct
+  let index_out_of_bounds f ~i ~length =
+    if length = 0 then
+      Printf.ksprintf invalid_arg
+        "Dynarray.%s: index %d out of bounds (empty dynarray)"
+        f i
+    else
+      Printf.ksprintf invalid_arg
+        "Dynarray.%s: index %d out of bounds (0..%d)"
+        f i (length - 1)
+
+  let negative_length f n =
+    Printf.ksprintf invalid_arg
+      "Dynarray.%s: negative length %d"
+      f n
+
+  let negative_capacity f n =
+    Printf.ksprintf invalid_arg
+      "Dynarray.%s: negative capacity %d"
+      f n
+
+  let requested_length_out_of_bounds f requested_length =
+    Printf.ksprintf invalid_arg
+      "Dynarray.%s: cannot grow to requested length %d (max_array_length is %d)"
+      f requested_length Sys.max_array_length
+
+  (* When observing an invalid state ([missing_element],
+     [invalid_length]), we do not give the name of the calling function
+     in the error message, as the error is related to invalid operations
+     performed earlier, and not to the callsite of the function
+     itself. *)
+
+  let missing_element ~i ~length =
+    Printf.ksprintf invalid_arg
+      "Dynarray: invalid array (missing element at position %d < length %d)"
+      i length
+
+  let invalid_length ~length ~capacity =
+    Printf.ksprintf invalid_arg
+      "Dynarray: invalid array (length %d > capacity %d)"
+      length capacity
+
+  let iterator_invalidation f ~expected ~observed =
+    Printf.ksprintf invalid_arg
+      "Dynarray.%s: iterator invalidated by a length change from %d to %d"
+      f expected observed
```

Comment:
The word "iterator" doesn't appear in the .mli, I don't know if this C++ term makes much sense here.

I think these scenarios warrant a new exception because the argument may become "invalid" long after the `iter` or whatever was called. Java seems to call it ConcurrentModification, C# uses InvalidOperation. `missing_element`, `invalid_length` make more sense as `Invalid_argument` (if they are checked at the beginning of a function), but they could be lumped with the new exception if they can be only due to concurrency. 

There is also the question of whether we really consider a dynarray an "invalid array" just because it changed during some iteration. If the internal state is fine, why would it be invalid? Concurrency is the issue, so I would call the exception something like `Dynarray.Concurrently_modified`, of course with the proviso that not all bad situations can be detected and the exception cannot be relied on.